src/EFCore.PG/Query/ExpressionVisitors/NpgsqlSqlTranslatingExpressionVisitor.cs

```diff
@@ -100,105 +69,117 @@ protected override Expression VisitBinary(BinaryExpression expression)
         /// An '= ANY' expression or null.
         /// </returns>
         [CanBeNull]
-        protected virtual Expression VisitEqualsAny([NotNull] SubQueryExpression expression)
-        {
-            var subQueryModel = expression.QueryModel;
-            var fromExpression = subQueryModel.MainFromClause.FromExpression;
+        protected override Expression VisitSubQuery(SubQueryExpression expression)
+            => base.VisitSubQuery(expression) ?? VisitArraySubQuery(expression);
 
-            var properties = MemberAccessBindingExpressionVisitor.GetPropertyPath(
-                fromExpression, _queryModelVisitor.QueryCompilationContext, out _);
+        /// <inheritdoc />
+        protected override Expression VisitUnary(UnaryExpression expression)
+        {
+            // Character literals are represented as integer literals by the expression tree.
+            // So `myString[0] == 'T'` looks like `((int)myString[0]) == 84`.
+            // Since `substr` returns `text`, not `char` or `int`, wrap it in `ascii`.
+            if (expression.NodeType == ExpressionType.Convert &&
+                Visit(expression.Operand) is IndexExpression index &&
+                index.Object.Type == typeof(string))
+                return new SqlFunctionExpression("ascii", typeof(int), new[] { index });
```

Comment:
Isn't the `substr` invocation missing? `ascii` only returns the code of the first character of the string...

Reply:
Nope, the substring is generated via the `IndexExpression`, so here we just have to make sure it's wrapped with `ascii` to handle the odd `character <-> integer` conversion. It looks a bit convoluted, but there are tests that verify the behavior.